### PR TITLE
Net dotnet washout

### PR DESCRIPTION
**Pull request by anjumjfax**

### CODE REVIEW

**Comment by anjumjfax** [comment]
An explanation of what I investigated, my findings, what I did.

**First attempt**

When I first started this task, it seemed quite as simple as modifying the current wash_out gem (0.11.0.biorails.30) to generate a WSDL with characterisics as described in this ticket; have it tack on a minOccurs=0 on all the attributes in the generator and call it a day. Manually editing the generated WSDL with these minOccurs and having .NET compile them actually did make the edited attributes a .NET Nullable<T> type. Things seemed hopeful, went ahead and modified the gem to do this everywhere.

Bad things ensued; the WSDL that was being generated thereafter was not valid, not to .NET or SOAP UI. Screwball errors appeared that whined about the schema not being valid and whatnot. Specifications for nullability differ whether it is an attribute in an object being passed up, passed back, and either case whether it is contained in a complex object or not.

I sunk the whole first week into attempting to solve this. It was like solving a rubiks cube, where I fix one side at the expense of messing up another side. I gave up, I need to go soul searching.

**Next attempt**

By the end of that week and into next week, I decided to try out upgrading to the upstream wash_out (0.12.0), a version that boasts rails 5 compatibility even. Made a few config changes to the controllers to make it compatible, and it ran. This version generated an extremely beautiful WSDL file, working flawlessly with SOAP UI. Better than ever even, because… every attribute was nullable. All the minOccurs and xsi:nillable were there in the right places. I was so damn hyped, everything looked up from here.

The sheer tragedy: It does not work with .NET. As one may observe from my previous comment, .NET gets real mad when you feed it the WSDL the new wash_out generates.

**The only option**

I must fix this new version of wash_out to work with .NET, preserving its capability to generate nullable attributes.

**The fixes**

_Basics_

First change I made to wash_out was to change the format of the names it gives to the end points generated. The names it gave contained ‘/' characters, so I just replaced them with '.’, and that actually fixed the error in the screenshot one sees in the previous comments.

_Date times._

Next I fixed now dates are presented in the SOAP response. .NET has no concept of dates, only datetimes, so made sure they are presented as datetimes at 00:00. More tricky than it sounds, involved recursing data structures, but it turned out as doable.

_Arrays_

Then it was about serializing back arrays contained in complex structures, like the Samples array in Container. Oh boy. How the SOAP response presented those arrays was not accepted. Turned out they needed to be wrapped in an tag saying they are an array, it’s length, needed an extra schema definition in the header, etc. Took me embarrassingly long to even find where the code is that does this, but I did it in the end.

_More arrays_

Then it was serializing back top level arrays that aren’t contained in complex structures, so methods like ‘lookups’ that return a direct array of ‘Lookup' objects. I don’t think this pattern is common SOAP, but it is accepted by .NET and figured out a way for to reimplement support for it. In the WSDL generator, I made every complex type have a counterpart special ‘array’ type, and made sure that special array type is the specified as the return type for the appropriate methods. That worked.

**There are still more issues**

I have left things there. In the course of all these fixes, it appears I have broken things in manner I haven’t characterised yet. Some api calls work, like login and container_get even, but for some reason lookups is broken (error reading what .NET sent up) and I haven’t dared test something like container_create.

Also, all the the type names in .NET are messed up. 'Core.Sessions' is 'Core.apiv5typesessions'. But posed the least of my worries atm.

**In conclusion**

I am done, for now. I have pushed these changes, pull request as ‘in progress’ is open on the biorails/wash_out repo. Yeah.
